design/continuous-delivery.md

```diff
@@ -0,0 +1,332 @@
+# Continuous Delivery for CDK Apps
+
+This is a specification for continuous delivery support for CDK apps. It
+describes the requirements, proposed APIs and developer workflow.
+
+**_Goal_: full CI/CD for CDK apps at any complexity.**
+
+The desired developer experience is that teams will be able to “git push” a
+change into their CDK app repo and have this change automatically picked up,
+built, tested and deployed according to a deployment flow they define.
+
+Any changes in resources, assets *or stacks* in their apps will automatically be
+added. To that end, the deployment pipeline itself will also be continuously
+delivered and updated throughout the same workflow. No manual deployments to
+production (incl. the pipeline) will be required.
+
+The only caveat is that **new environments** (account/region) will need to be
+bootstrapped in advance in order to establish trust with the central pipeline
+environment and set-up resources needed for deployment such as the assets S3
+bucket.
+
+Prototype: https://github.com/eladb/cdkcd-test
+
+## Requirements
+
+This list describes only the minimal set of requirements from this feature.
+After we release these building blocks, we will look into vending higher-level
+"one liner" APIs that will make it very easy to get started.
+
+1. **Assets**: Support apps that include all supported kinds of assets (S3
+   files, ECR images)
+2. **Multiple-environments**: Support apps that have stacks that target multiple
+   environments (accounts/regions)
+3. **Orchestration**: Allow developers to express complex deployment
+   orchestration based on the capabilities of CodePipeline
+4. **User defined build + synth environment**: the runtime environment in which
+   the code is built and the CDK app is synthesized should be fully customizable
+   by the user
+5. **Controlled deployment environment**: for security reasons, the runtime
+   environment in which deployment is executed will be fully controlled and will
+   not allow running user code or user-defined image
+
+_Non-requirements/assumptions:_
+
+1. Bundling assets (creating Lambda zip files, building docker images) and
+   publishing them (upload to S3, push to ECR) is currently coupled with the
+   deployment of each stack that uses them. This means, for example, that if a
+   docker image is used in 3 stacks, it will be built 3 times, and this could
+   lead to different images (for example, “apt get install” can install a
+   different version of a dependency each time). We are assuming that this is
+   good enough for the first version of this feature.
+2. We assume that **cdk.context.json** is committed into the repo. Any
+   context-fetching will be done manually by users and committed to the
+   repository.
+3. There’s a one-to-one mapping between an app and a pipeline. We are not
```

Comment:
Will there be a configurability for the pipeline? for eg: different apps with different permissions model may follow the same pipeline structure.
eg: deploy pipeline of format 'A' for app 'A' and app 'B'

Reply:
Yes, but this is something we need to add to the codepipelines library in general (regardless of the CDK). @skinny85 ==